extensions/catalog/federated-catalog-cache/src/main/java/org/eclipse/dataspaceconnector/catalog/cache/FederatedCatalogCacheExtension.java

```diff
@@ -0,0 +1,186 @@
+        var queryEngine = new QueryEngineImpl(queryAdapterRegistry);
+        context.registerService(QueryEngine.class, queryEngine);
+        monitor = context.getMonitor();
+        var catalogController = new CatalogController(monitor, queryEngine);
+        webService.registerController(catalogController);
+
+        // CRAWLER SUBSYSTEM
+        updateResponseQueue = new ArrayBlockingQueue<>(DEFAULT_QUEUE_LENGTH);
+
+        //todo: maybe get this from a database or somewhere else?
+        partitionManagerConfig = new PartitionConfiguration(context);
+
+        // lets create a simple partition manager
+        partitionManager = createPartitionManager(context, updateResponseQueue);
+
+        // and a loader manager
+        loaderManager = createLoaderManager(store);
+
+        monitor.info("Federated Catalog Cache extension initialized");
+    }
+
+    @Override
+    public void start() {
+        partitionManager.schedule(partitionManagerConfig.getExecutionPlan());
+        loaderManager.start(updateResponseQueue);
+        monitor.info("Federated Catalog Cache extension started");
+    }
+
+    @Override
+    public void shutdown() {
+        partitionManager.stop();
+        loaderManager.stop();
+        monitor.info("Federated Catalog Cache extension stopped");
+    }
+
+    @NotNull
+    private LoaderManager createLoaderManager(FederatedCacheStore store) {
+        return new LoaderManagerImpl(List.of(createLoader(store)),
+                partitionManagerConfig.getLoaderBatchSize(DEFAULT_BATCH_SIZE),
+                () -> partitionManagerConfig.getLoaderRetryTimeout(DEFAULT_RETRY_TIMEOUT_MILLIS), monitor);
+    }
+
+    @NotNull
+    private PartitionManager createPartitionManager(ServiceExtensionContext context, ArrayBlockingQueue<UpdateResponse> updateResponseQueue) {
+
+        // protocol registry - must be supplied by another extension
+        var protocolAdapterRegistry = context.getService(CatalogQueryAdapterRegistry.class);
+
+        // get all known nodes from node directory - must be supplied by another extension
+        var directory = context.getService(FederatedCacheNodeDirectory.class);
+
+        // use all nodes EXCEPT self
+        Supplier<List<WorkItem>> nodes = () -> directory.getAll().stream()
+                .filter(node -> !node.getName().equals(context.getConnectorId()))
+                .map(n -> new WorkItem(n.getTargetUrl(), selectProtocol(n.getSupportedProtocols()))).collect(Collectors.toList());
+
+        return new PartitionManagerImpl(monitor,
+                new DefaultWorkItemQueue(partitionManagerConfig.getWorkItemQueueSize(10)),
+                workItems -> createCrawler(workItems, context, protocolAdapterRegistry, updateResponseQueue),
+                partitionManagerConfig.getNumCrawlers(DEFAULT_NUM_CRAWLERS),
+                nodes);
+    }
+
+    @NotNull
+    private Loader createLoader(FederatedCacheStore store) {
+        return responses -> {
```

Comment:
i moved it to a separate class. I didn't find a more convenient implementation using `flatMap`, because we always have _one_ originator plus _several_ assets. Am I missing something?

Reply:
Just matter of taste, for me will work also like this.
Just another question on the naming of the `Loader` implementation: I saw you used the `Default*` convention (which I prefer, honestly), but elsewhere in the codebase the `*Impl` convention is used, maybe should be better to mantain this convention and, why not, report it on the documentation.